nexus/src/authz/context.rs

```diff
@@ -63,49 +77,80 @@ impl Context {
         resource: Resource,
     ) -> Result<(), Error>
     where
-        Resource: oso::ToPolar + AuthzResource,
+        Resource: oso::ToPolar + Authorize + Clone,
     {
-        // TODO-security For Action::Read (and any other "read" action),
-        // this should return NotFound rather than Forbidden.  But we cannot
-        // construct the appropriate NotFound here without more information:
-        // the resource type and how it was looked up.  In practice, it's
-        // quite possible that all such cases should really use query
-        // filtering instead of an explicit is_allowed() check, in which
-        // case we could safely assume Forbidden here.
-        //
-        // Alternatively, we could let the caller produce the appropriate
-        // "NotFound", but it would add a lot of boilerplate to a lot of
-        // callers if we didn't return api::external::Error here.
         let mut roles = RoleSet::new();
         resource
-            .fetch_all_related_roles_for_user(
-                opctx,
-                &self.datastore,
-                &self.authn,
-                &mut roles,
-            )
+            .load_roles(opctx, &self.datastore, &self.authn, &mut roles)
             .await?;
         debug!(opctx.log, "roles"; "roles" => ?roles);
         let actor = AnyActor::new(&self.authn, roles);
         let is_authn = self.authn.actor().is_some();
-        match self.authz.oso.is_allowed(actor, action, resource) {
+        match self.authz.is_allowed(&actor, action, &resource) {
+            Ok(true) => Ok(()),
             Err(error) => Err(Error::internal_error(&format!(
                 "failed to compute authorization: {:#}",
                 error
             ))),
-            // If the user did not authenticate successfully, this will become a
-            // 401 rather than a 403.
-            Ok(false) if !is_authn => Err(Error::Unauthenticated {
-                internal_message: String::from(
-                    "authorization failed for unauthenticated request",
-                ),
-            }),
-            Ok(false) => Err(Error::Forbidden),
-            Ok(true) => Ok(()),
+            Ok(false) => {
+                let error = if is_authn {
+                    Error::Forbidden
+                } else {
+                    // If the user did not authenticate successfully, this will
+                    // become a 401 rather than a 403.
+                    Error::Unauthenticated {
+                        internal_message: String::from(
+                            "authorization failed for unauthenticated request",
+                        ),
+                    }
+                };
+
+                resource.on_unauthorized(&self.authz, error, actor, action)
```

Comment:
Why is the logic in `on_unauthorized` part of the trait rather than inlined here? Do you picture there being multiple implementations? It feels ok as-is, but I find the relationship between the function and the calling context a little confusing, in a way that it wouldn't be if it was inlined. The `on_` naming helps a lot, though.

Reply:
Just kidding, I see there already are multiple implementations.